examples/c++-api/AtomSpaceEventSubscribeExample.cc

```diff
@@ -0,0 +1,125 @@
+#include <iostream>
+
+#include <boost/signals2.hpp>
+
+#include <opencog/atomspace/AtomSpace.h>
+#include <opencog/truthvalue/AttentionValue.h>
+#include <opencog/truthvalue/SimpleTruthValue.h>
+#include <opencog/truthvalue/TruthValue.h>
+
+using namespace opencog;
+
+/**
+ * Sample code that shows how to listen for AtomSpace related events.
+ * The followings are all the events/signals supported by the AtomSpace.
+ */
+
+void AtomAddedCBHandler(const Handle& h);
+void AVChangedCBHandler(const Handle& h, const AttentionValuePtr& av_old,
+                        const AttentionValuePtr& av_new);
+void TVChangedCBHandler(const Handle& h, const TruthValuePtr& av_old,
+                        const TruthValuePtr& tv_new);
+void AtomRemovedCBHandler(const AtomPtr&);
+void AtomAddedToAFCBHandler(const Handle& h, const AttentionValuePtr& av_old,
+                            const AttentionValuePtr& av_new);
+void AtomRemovedFromAFCBHandler(const Handle& h,
+                                const AttentionValuePtr& av_old,
+                                const AttentionValuePtr& av_new);
+
+int main(int argc, char ** args)
+{
+    AtomSpace as;
+
+    boost::signals2::connection AtomAddedSignalConn,
+                                AVChangedSignalConn,
+                                TVChangedSignalConn,
+                                AtomsRemovedSignalConn,
+                                AtomAddedToAttentionalFocusSignalConn,
+                                AtomRemovedFromAttentionalFocusSignalConn;
+
+    //Register Atomspace Event call back handlers.
+    AtomAddedSignalConn = as.addAtomSignal(
+            boost::bind(&AtomAddedCBHandler, _1));
+    AVChangedSignalConn = as.AVChangedSignal(
+            boost::bind(&AVChangedCBHandler, _1, _2, _3));
+    TVChangedSignalConn = as.TVChangedSignal(
+            boost::bind(&TVChangedCBHandler, _1, _2, _3));
+    AtomsRemovedSignalConn = as.removeAtomSignal(
+            boost::bind(&AtomRemovedCBHandler, _1));
+    AtomAddedToAttentionalFocusSignalConn = as.AddAFSignal(
+            boost::bind(&AtomAddedToAFCBHandler, _1, _2, _3));
+    AtomRemovedFromAttentionalFocusSignalConn = as.RemoveAFSignal(
+            boost::bind(&AtomRemovedFromAFCBHandler, _1, _2, _3));
```

Comment:
I am very very concerned about the addition of this example, for many reasons.  It encourages code debt!  Code debt is bad!

https://en.wikipedia.org/wiki/Technical_debt

1)  Most programmers should not be using the C++ at all.
2) The performance implications of the atomspace event signals are disasterous!  These signals should not be used, if at all possible.
3) We should not be writing examples for an API whose use we want to discourage!  
4) It is very easy to write code for an API, it is much much harder to remove code for an API. If someone starts to use this API, it will be quite hard to remove: someone ill have to do the work to redesign the code!!  Who is going to do that work?   It is much better to simpl not create the code to begin with, than too start by incurring a huge liability.

Basically, dont go into code debt on day one.  Try to avoid code debt.  These examples will encourage debt! 

 https://en.wikipedia.org/wiki/Technical_debt